### PR TITLE
bpo-1635741: Port gc module to multiphase initialization

### PR DESCRIPTION
**Pull request by tiran**
Signed-off-by: Christian Heimes <christian@python.org>

https://bugs.python.org/issue1635741

Automerge-Triggered-By: GH:tiran

### CODE REVIEW

**Comment by pablogsal** [comment]
LGTM, but I added the buildbot label because the gc module has some special handling on deallocation

**Comment by tiran** [comment]
refleak bots are complaining about a new reference like in ``test_ast``.

**Comment by tiran** [comment]
> refleak bots are complaining about a new reference like in `test_ast`.

Found it! ``gcstate->garbage`` is initialized much earlier at the beginning of an interpreter lifecycle.

File: Modules/gcmodule.c [hunk 1]
```diff
@@ -1992,59 +1992,58 @@ static PyMethodDef GcMethods[] = {
...
     }
-    Py_INCREF(gcstate->garbage);
-    if (PyModule_AddObject(m, "garbage", gcstate->garbage) < 0) {
-        return NULL;
+    if (PyModule_AddObjectRef(module, "garbage", gcstate->garbage) < 0) {
+        return -1;
     }
 
+    assert(gcstate->callbacks == NULL);
+    gcstate->callbacks = PyList_New(0);
```
Comment (vstinner) [inline Modules/gcmodule.c]: This looks wrong to me. If you create multiple instances of the GC modul, callbacks member will be overriden by a new list at each call. It must be initialized exactly once in _PyGC_Init().

By the way, `if (gcstate->garbage == NULL) {` test in _PyGC_Init() can be removed, it's useless. This function is called exactly once per interpreter.

File: Modules/gcmodule.c [hunk 2]
```diff
@@ -1992,59 +1992,58 @@ static PyMethodDef GcMethods[] = {
...
     GCState *gcstate = get_gc_state();
 
-    PyObject *m = PyModule_Create(&gcmodule);
-
-    if (m == NULL) {
-        return NULL;
-    }
-
+    /* Initialized by _PyGC_Init() early in interpreter lifecycle */
     if (gcstate->garbage == NULL) {
```
Comment (vstinner) [inline Modules/gcmodule.c]: This test is useless and can be removed. You cannot import the gc module if _PyGC_Init() failed. If _PyGC_Init() fails, Python exits immediately. You can remove the whole test.

I probably forgot to remove it when I wrote _PyGC_Init().

File: Modules/gcmodule.c [hunk 3]
```diff
@@ -1992,59 +1992,58 @@ static PyMethodDef GcMethods[] = {
...
+static PyModuleDef_Slot gcmodule_slots[] = {
+    {Py_mod_exec, gcmodule_exec},
+    {0, NULL}
+};
+
+static struct PyModuleDef gcmodule = {
+    PyModuleDef_HEAD_INIT,
+    .m_name = "gc",
+    .m_doc = gc__doc__,
+    .m_size = 0,  /* special case, state is part of interpreter state */
```
Comment (vstinner) [inline Modules/gcmodule.c]: Would you mind to mention get_gc_state()? I suggest:

```suggestion
    .m_size = 0,  // Per interpreter state, see: get_gc_state()
```

File: Modules/gcmodule.c (hunk 1 above)
Comment (tiran) [reply Modules/gcmodule.c]: Makes sense, I'm moving ``gcstate->callbacks = PyList_New(0)`` to ``_PyGC_Init()``.

File: Modules/gcmodule.c (hunk 3 above)
Comment (tiran) [reply Modules/gcmodule.c]: Sure

**Review by vstinner — APPROVED** [review]
LGTM.

**Pull request merged**